Make TCP connect handle EINTR correctly

darthunix opened the pull request:
According to the [POSIX](https://pubs.opengroup.org/onlinepubs/009695399/functions/connect.html) standard, if connect() is interrupted by a signal that is caught while blocked waiting to establish a connection, connect() shall fail and set errno to EINTR, but the connection request shall not be aborted, and the connection shall be established asynchronously. When the connection has been established asynchronously, select() and poll() shall indicate that the file descriptor for the socket is ready for writing.

The previous implementation differs from the recomendation: in a case of the EINTR we tried to reconnect in a loop and sometimes get EISCONN error (this problem was originally detected on MacOS).

1. More details about the problem in an [article](http://www.madore.org/~david/computers/connect-intr.html).
2. The original [issue](https://git.picodata.io/picodata/picodata/tarantool-module/-/issues/157).

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6294367807/job/17086110884) [(plain)](https://github.com/rust-lang/rust/commit/4be1dfbefbf6193b140a1dd3f0da84ffba952d8d/checks/17086110884/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=darthunix
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_2d3afaa8-38b3-4a6b-a1e6-0dae6a4b2acd
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=301c9e86a69ef95be18e6499c5ddd95957ca0f09
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_2d3afaa8-38b3-4a6b-a1e6-0dae6a4b2acd
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_2d3afaa8-38b3-4a6b-a1e6-0dae6a4b2acd
GITHUB_TRIGGERING_ACTOR=darthunix
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116132/merge
GITHUB_WORKFLOW_SHA=301c9e86a69ef95be18e6499c5ddd95957ca0f09
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
    Checking addr2line v0.21.0
error[E0422]: cannot find struct, variant or union type `pollfd` in crate `libc`
   --> library/std/src/sys_common/net.rs:239:36
    |
239 |             let mut pollfd = libc::pollfd { fd: sock.as_raw(), events: libc::POLLOUT, revents: 0 };


error[E0425]: cannot find value `POLLOUT` in crate `libc`
    |
    |
239 |             let mut pollfd = libc::pollfd { fd: sock.as_raw(), events: libc::POLLOUT, revents: 0 };


error[E0425]: cannot find function `poll` in crate `libc`
    |
    |
241 |                 let ret = unsafe { libc::poll(&mut pollfd, 1, -1) };

Some errors have detailed explanations: E0422, E0425.
For more information about an error, try `rustc --explain E0422`.
error: could not compile `std` (lib) due to 3 previous errors
```

</details>

darthunix commented:
After some research (linux/darwin-xnu sources, golang [connect](https://github.com/golang/go/blob/0c64ebce7e7007615aba58a19be07372f0febe69/src/net/fd_unix.go#L55)) I think it is safe to call `connect()` in a loop as we do now. The only thing we should do is to handle EISCONN error when `connect()` was interrupted by a signal and connected asynchronously before the next loop iteration (as it often happens on MacOS).

the8472 commented:
You shouldn't change `cvt_r` though since it's also used in other places which don't care about `EISCONN`.

darthunix committed:
Make TCP connect() handle EINTR correctly

According to the POSIX standard, if connect() is interrupted by a
signal that is caught while blocked waiting to establish a connection,
connect() shall fail and set errno to EINTR, but the connection
request shall not be aborted, and the connection shall be established
asynchronously.

If asynchronous connection was successfully established after EINTR
and before the next connection attempt, OS returns EISCONN that was
handled as an error before. This behavior is fixed now and we handle
it as success.

The problem affects MacOS users: Linux doesn't return EISCONN in this
case, Windows connect() can not be interrupted without an old-fashoin
WSACancelBlockingCall function that is not used in the library.
So current solution gives connect() as OS specific implementation.

darthunix commented:
@the8472 can you take a look at the current implementation? I have no experience with hermit and soild os, so I've left everything as it (for solid os `cvt_r` is equal to `cvt`, so I've made a small refactoring)

workingjubilee commented:
cc @kawadakk for Solid
cc @stlankes @mkroening for Hermit

mkroening commented on library/std/src/sys/hermit/net.rs:
```diff
@@ -56,6 +56,11 @@ impl Socket {
 
+    pub fn connect(&self, addr: &SocketAddr) -> io::Result<()> {
+        let (addr, len) = addr.into_inner();
+        cvt_r(unsafe { netc::connect(self.as_raw_fd(), addr.as_ptr(), len) })
```
This does not compile.

```suggestion
        cvt_r(|| unsafe { netc::connect(self.as_raw_fd(), addr.as_ptr(), len) })?;
        Ok(())
```

darthunix replied:
Thanks for this catch. I have originally used `cvt` instead of the `cvt_r` here, but then realized that I don't know anything about hermit os and whether it can interrupt connect by a signal or not. So, finally I decided to leave everything as it was before. BTW, is there any documentation to read about connect on hermit?

mkroening replied:
> is there any documentation to read about connect on hermit?

I am afraid not. There's the function definition in [`hermit_abi::connect`](https://docs.rs/hermit-abi/latest/hermit_abi/fn.connect.html) and there is [the actual implementation](https://github.com/search?q=repo%3Ahermit-os%2Fkernel+connect&type=code). It's not documented, though, and may change in the future. Currently, we mimic BSD sockets.

mkroening reviewed (changes requested):
Thanks for the notification! :)

The Hermit changes do not compile.

mkroening reviewed (approved):
Compiles for Hermit now, thanks. :)

kawadakk commented on library/std/src/sys/solid/net.rs:
```diff
@@ -233,12 +233,14 @@ impl Socket {
 
+    pub fn connect(&self, addr: &SocketAddr) -> io::Result<()> {
+        let (addr, len) = addr.into_inner();
+        cvt(unsafe { netc::connect(self.0.raw(), addr.as_ptr(), len) })
```
```suggestion
        cvt(unsafe { netc::connect(self.0.raw(), addr.as_ptr(), len) })?;
        Ok(())
```

darthunix replied:
Done!

kawadakk reviewed (changes requested):
Thank you for heads-up!

kawadakk reviewed (approved):
This compiles for SOLID targets now. Thanks!

cuviper commented:
@bors r+

bors commented:
:pushpin: Commit dfadd177a9738de289a3a6d4b39c08d94f384c1a has been approved by `cuviper`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cuviper dfadd177a9738de289a3a6d4b39c08d94f384c1a -->
<!-- homu: {"type":"Approved","sha":"dfadd177a9738de289a3a6d4b39c08d94f384c1a","approver":"cuviper","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit dfadd177a9738de289a3a6d4b39c08d94f384c1a with merge 3fbcfd2b6f7030cb70328aa759107efc1516912c...
<!-- homu: {"type":"BuildStarted","head_sha":"dfadd177a9738de289a3a6d4b39c08d94f384c1a","merge_sha":"3fbcfd2b6f7030cb70328aa759107efc1516912c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6573860783/job/17862022947)
Approved by: cuviper
Pushing 3fbcfd2b6f7030cb70328aa759107efc1516912c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cuviper","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6573860783/job/17862022947"},"merge_sha":"3fbcfd2b6f7030cb70328aa759107efc1516912c"} -->
